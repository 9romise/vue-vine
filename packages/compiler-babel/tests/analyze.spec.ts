import { describe, expect, test } from 'vitest'
import { compileVineTypeScriptFile } from '../index'
import { createMockTransformCtx } from './shared-utils'

describe('Test Vine compiler analyze', () => {
  test('analyze vine component props by function\'s formal param', () => {
    const content = `
function MyComp(p: {
  name: string;
  data: SomeExternalType;
  bool: boolean;
}) {
  return vine\`<div>Test props by formal param</div>\`
}`
    const { mockCompilerCtx, mockCompilerHooks } = createMockTransformCtx()
    compileVineTypeScriptFile(content, 'testAnalyzeVinePropsByFormalParam', mockCompilerHooks)
    expect(mockCompilerCtx.vineCompileErrors.length).toBe(0)
    expect(mockCompilerCtx.fileCtxMap.size).toBe(1)
    const fileCtx = mockCompilerCtx.fileCtxMap.get('testAnalyzeVinePropsByFormalParam')
    expect(fileCtx?.vineFnComps.length).toBe(1)
    const vineFnComp = fileCtx?.vineFnComps[0]
    expect(vineFnComp?.propsAlias).toBe('p')
    expect(vineFnComp?.props).toMatchInlineSnapshot(`
      {
        "bool": {
          "isBool": true,
          "isFromMacroDefine": false,
          "isRequired": true,
        },
        "data": {
          "isBool": false,
          "isFromMacroDefine": false,
          "isRequired": true,
        },
        "name": {
          "isBool": false,
          "isFromMacroDefine": false,
          "isRequired": true,
        },
      }
    `)
  })

  test('analyze vine component props by macro calls', () => {
    const content = `
const MyComp = () => {
  const name = vineProp<string>()
  const disabled = vineProp.optional<boolean>()
  const title = vineProp.withDefault('# Title', (val: string) => val.startsWith('#'))
  return vine\`<div>Test props by macro calls</div>\`
}`
    const { mockCompilerCtx, mockCompilerHooks } = createMockTransformCtx()
    compileVineTypeScriptFile(content, 'testAnalyzeVinePropsByMacroCalls', mockCompilerHooks)
    expect(mockCompilerCtx.vineCompileErrors.length).toBe(0)
    expect(mockCompilerCtx.fileCtxMap.size).toBe(1)
    const fileCtx = mockCompilerCtx.fileCtxMap.get('testAnalyzeVinePropsByMacroCalls')
    expect(fileCtx?.vineFnComps.length).toBe(1)
    const vineFnComp = fileCtx?.vineFnComps[0]
    expect(vineFnComp?.props).toMatchSnapshot()
  })

  test('analyze vine emits definition', () => {
    const content = `
function MyComp() {
  const myEmits = vineEmits<{
    foo: (a: string) => void;
    bar: (b: number) => void;
  }>()
  return vine\`<div>Test emits</div>\`
}`
    const { mockCompilerCtx, mockCompilerHooks } = createMockTransformCtx()
    compileVineTypeScriptFile(content, 'testAnalyzeVineEmits', mockCompilerHooks)
    expect(mockCompilerCtx.vineCompileErrors.length).toBe(0)
    expect(mockCompilerCtx.fileCtxMap.size).toBe(1)
    const fileCtx = mockCompilerCtx.fileCtxMap.get('testAnalyzeVineEmits')
    expect(fileCtx?.vineFnComps.length).toBe(1)
    const vineFnComp = fileCtx?.vineFnComps[0]
    expect(vineFnComp?.emitsAlias).toBe('myEmits')
    expect(vineFnComp?.emits).toEqual(['foo', 'bar'])
  })

  test('analyze `vineExpose` and `vineOptions` macro calls', () => {
    const content = `
function Comp() {
  const count = ref(1)
  vineExpose({ count })
  vineOptions({
    name: 'MyComp',
    inheritAttrs: false,
  })
  return vine\`<div>Test expose and options</div>\`
}`
    const { mockCompilerCtx, mockCompilerHooks } = createMockTransformCtx()
    compileVineTypeScriptFile(content, 'testAnalyzeVineExposeAndOptions', mockCompilerHooks)
    expect(mockCompilerCtx.vineCompileErrors.length).toBe(0)
    expect(mockCompilerCtx.fileCtxMap.size).toBe(1)
    const fileCtx = mockCompilerCtx.fileCtxMap.get('testAnalyzeVineExposeAndOptions')
    expect(fileCtx?.vineFnComps.length).toBe(1)
    const vineFnComp = fileCtx?.vineFnComps[0]
    expect(vineFnComp?.expose).toMatchSnapshot()
    expect(vineFnComp?.options).toMatchSnapshot()
  })
})
