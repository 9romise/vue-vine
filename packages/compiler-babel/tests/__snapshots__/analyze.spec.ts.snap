// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test Vine compiler analyze > analyze \`vineExpose\` and \`vineOptions\` macro calls 1`] = `
Node {
  "end": 64,
  "loc": SourceLocation {
    "end": Position {
      "column": 22,
      "index": 64,
      "line": 4,
    },
    "filename": undefined,
    "identifierName": undefined,
    "start": Position {
      "column": 13,
      "index": 55,
      "line": 4,
    },
  },
  "properties": [
    Node {
      "computed": false,
      "end": 62,
      "extra": {
        "shorthand": true,
      },
      "key": Node {
        "end": 62,
        "loc": SourceLocation {
          "end": Position {
            "column": 20,
            "index": 62,
            "line": 4,
          },
          "filename": undefined,
          "identifierName": "count",
          "start": Position {
            "column": 15,
            "index": 57,
            "line": 4,
          },
        },
        "name": "count",
        "start": 57,
        "type": "Identifier",
      },
      "loc": SourceLocation {
        "end": Position {
          "column": 20,
          "index": 62,
          "line": 4,
        },
        "filename": undefined,
        "identifierName": undefined,
        "start": Position {
          "column": 15,
          "index": 57,
          "line": 4,
        },
      },
      "method": false,
      "shorthand": true,
      "start": 57,
      "type": "ObjectProperty",
      "value": Node {
        "end": 62,
        "extra": undefined,
        "loc": SourceLocation {
          "end": Position {
            "column": 20,
            "index": 62,
            "line": 4,
          },
          "filename": undefined,
          "identifierName": "count",
          "start": Position {
            "column": 15,
            "index": 57,
            "line": 4,
          },
        },
        "name": "count",
        "range": undefined,
        "start": 57,
        "type": "Identifier",
      },
    },
  ],
  "start": 55,
  "type": "ObjectExpression",
}
`;

exports[`Test Vine compiler analyze > analyze \`vineExpose\` and \`vineOptions\` macro calls 2`] = `
Node {
  "end": 130,
  "extra": {
    "trailingComma": 125,
  },
  "loc": SourceLocation {
    "end": Position {
      "column": 3,
      "index": 130,
      "line": 8,
    },
    "filename": undefined,
    "identifierName": undefined,
    "start": Position {
      "column": 14,
      "index": 80,
      "line": 5,
    },
  },
  "properties": [
    Node {
      "computed": false,
      "end": 100,
      "key": Node {
        "end": 90,
        "loc": SourceLocation {
          "end": Position {
            "column": 8,
            "index": 90,
            "line": 6,
          },
          "filename": undefined,
          "identifierName": "name",
          "start": Position {
            "column": 4,
            "index": 86,
            "line": 6,
          },
        },
        "name": "name",
        "start": 86,
        "type": "Identifier",
      },
      "loc": SourceLocation {
        "end": Position {
          "column": 18,
          "index": 100,
          "line": 6,
        },
        "filename": undefined,
        "identifierName": undefined,
        "start": Position {
          "column": 4,
          "index": 86,
          "line": 6,
        },
      },
      "method": false,
      "shorthand": false,
      "start": 86,
      "type": "ObjectProperty",
      "value": Node {
        "end": 100,
        "extra": {
          "raw": "'MyComp'",
          "rawValue": "MyComp",
        },
        "loc": SourceLocation {
          "end": Position {
            "column": 18,
            "index": 100,
            "line": 6,
          },
          "filename": undefined,
          "identifierName": undefined,
          "start": Position {
            "column": 10,
            "index": 92,
            "line": 6,
          },
        },
        "start": 92,
        "type": "StringLiteral",
        "value": "MyComp",
      },
    },
    Node {
      "computed": false,
      "end": 125,
      "key": Node {
        "end": 118,
        "loc": SourceLocation {
          "end": Position {
            "column": 16,
            "index": 118,
            "line": 7,
          },
          "filename": undefined,
          "identifierName": "inheritAttrs",
          "start": Position {
            "column": 4,
            "index": 106,
            "line": 7,
          },
        },
        "name": "inheritAttrs",
        "start": 106,
        "type": "Identifier",
      },
      "loc": SourceLocation {
        "end": Position {
          "column": 23,
          "index": 125,
          "line": 7,
        },
        "filename": undefined,
        "identifierName": undefined,
        "start": Position {
          "column": 4,
          "index": 106,
          "line": 7,
        },
      },
      "method": false,
      "shorthand": false,
      "start": 106,
      "type": "ObjectProperty",
      "value": Node {
        "end": 125,
        "loc": SourceLocation {
          "end": Position {
            "column": 23,
            "index": 125,
            "line": 7,
          },
          "filename": undefined,
          "identifierName": undefined,
          "start": Position {
            "column": 18,
            "index": 120,
            "line": 7,
          },
        },
        "start": 120,
        "type": "BooleanLiteral",
        "value": false,
      },
    },
  ],
  "start": 80,
  "type": "ObjectExpression",
}
`;

exports[`Test Vine compiler analyze > analyze vine component props by macro calls 1`] = `
{
  "disabled": {
    "isBool": true,
    "isFromMacroDefine": true,
    "isRequired": false,
    "validator": undefined,
  },
  "name": {
    "isBool": false,
    "isFromMacroDefine": true,
    "isRequired": true,
    "validator": undefined,
  },
  "title": {
    "default": Node {
      "end": 152,
      "extra": {
        "raw": "'# Title'",
        "rawValue": "# Title",
      },
      "loc": SourceLocation {
        "end": Position {
          "column": 46,
          "index": 152,
          "line": 5,
        },
        "filename": undefined,
        "identifierName": undefined,
        "start": Position {
          "column": 37,
          "index": 143,
          "line": 5,
        },
      },
      "start": 143,
      "type": "StringLiteral",
      "value": "# Title",
    },
    "isBool": false,
    "isFromMacroDefine": true,
    "isRequired": true,
    "validator": Node {
      "async": false,
      "body": Node {
        "arguments": [
          Node {
            "end": 189,
            "extra": {
              "raw": "'#'",
              "rawValue": "#",
            },
            "loc": SourceLocation {
              "end": Position {
                "column": 83,
                "index": 189,
                "line": 5,
              },
              "filename": undefined,
              "identifierName": undefined,
              "start": Position {
                "column": 80,
                "index": 186,
                "line": 5,
              },
            },
            "start": 186,
            "type": "StringLiteral",
            "value": "#",
          },
        ],
        "callee": Node {
          "computed": false,
          "end": 185,
          "loc": SourceLocation {
            "end": Position {
              "column": 79,
              "index": 185,
              "line": 5,
            },
            "filename": undefined,
            "identifierName": undefined,
            "start": Position {
              "column": 65,
              "index": 171,
              "line": 5,
            },
          },
          "object": Node {
            "end": 174,
            "loc": SourceLocation {
              "end": Position {
                "column": 68,
                "index": 174,
                "line": 5,
              },
              "filename": undefined,
              "identifierName": "val",
              "start": Position {
                "column": 65,
                "index": 171,
                "line": 5,
              },
            },
            "name": "val",
            "start": 171,
            "type": "Identifier",
          },
          "property": Node {
            "end": 185,
            "loc": SourceLocation {
              "end": Position {
                "column": 79,
                "index": 185,
                "line": 5,
              },
              "filename": undefined,
              "identifierName": "startsWith",
              "start": Position {
                "column": 69,
                "index": 175,
                "line": 5,
              },
            },
            "name": "startsWith",
            "start": 175,
            "type": "Identifier",
          },
          "start": 171,
          "type": "MemberExpression",
        },
        "end": 190,
        "loc": SourceLocation {
          "end": Position {
            "column": 84,
            "index": 190,
            "line": 5,
          },
          "filename": undefined,
          "identifierName": undefined,
          "start": Position {
            "column": 65,
            "index": 171,
            "line": 5,
          },
        },
        "start": 171,
        "type": "CallExpression",
      },
      "end": 190,
      "generator": false,
      "id": null,
      "loc": SourceLocation {
        "end": Position {
          "column": 84,
          "index": 190,
          "line": 5,
        },
        "filename": undefined,
        "identifierName": undefined,
        "start": Position {
          "column": 48,
          "index": 154,
          "line": 5,
        },
      },
      "params": [
        Node {
          "end": 166,
          "loc": SourceLocation {
            "end": Position {
              "column": 60,
              "index": 166,
              "line": 5,
            },
            "filename": undefined,
            "identifierName": "val",
            "start": Position {
              "column": 49,
              "index": 155,
              "line": 5,
            },
          },
          "name": "val",
          "start": 155,
          "type": "Identifier",
          "typeAnnotation": Node {
            "end": 166,
            "loc": SourceLocation {
              "end": Position {
                "column": 60,
                "index": 166,
                "line": 5,
              },
              "filename": undefined,
              "identifierName": undefined,
              "start": Position {
                "column": 52,
                "index": 158,
                "line": 5,
              },
            },
            "start": 158,
            "type": "TSTypeAnnotation",
            "typeAnnotation": Node {
              "end": 166,
              "loc": SourceLocation {
                "end": Position {
                  "column": 60,
                  "index": 166,
                  "line": 5,
                },
                "filename": undefined,
                "identifierName": undefined,
                "start": Position {
                  "column": 54,
                  "index": 160,
                  "line": 5,
                },
              },
              "start": 160,
              "type": "TSStringKeyword",
            },
          },
        },
      ],
      "start": 154,
      "type": "ArrowFunctionExpression",
    },
  },
}
`;
